File: WdtcCore/src/main/java/org/wdt/wdtc/core/download/game/DownloadGameAssetsFile.java
package org.wdt.wdtc.core.download.game;

import com.google.gson.JsonElement;
import com.google.gson.annotations.SerializedName;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.SneakyThrows;
import org.apache.log4j.Logger;
import org.wdt.utils.gson.JsonObjectUtils;
import org.wdt.utils.gson.JsonUtils;
import org.wdt.utils.io.FileUtils;
import org.wdt.wdtc.core.download.SpeedOfProgress;
import org.wdt.wdtc.core.download.infterface.DownloadSourceInterface;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.manger.DownloadSourceManger;
import org.wdt.wdtc.core.utils.DownloadUtils;
import org.wdt.wdtc.core.utils.ThreadUtils;
import org.wdt.wdtc.core.utils.WdtcLogger;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.util.Map;

public class DownloadGameAssetsFile {
  private static final Logger logmaker = WdtcLogger.getLogger(DownloadGameAssetsFileTask.class);
  protected final Launcher launcher;

  public DownloadGameAssetsFile(Launcher launcher) {
    this.launcher = launcher;
  }

  @SneakyThrows(IOException.class)
  public void DownloadAssetsFiles() {
    Map<String, JsonElement> list = JsonUtils.readFileToJsonObject(launcher.getGameAssetsListJson()).getAsJsonObject("objects").asMap();
    SpeedOfProgress progress = new SpeedOfProgress(list.size());
    for (String key : list.keySet()) {
      AssetsFileData data = JsonObjectUtils.parseObject(list.get(key).getAsJsonObject(), AssetsFileData.class);
      if (DownloadUtils.isDownloadProcess()) return;
      if (FileUtils.isFileNotExistsAndIsNotSameSize(new File(launcher.getGameObjects(), data.getHashSplicing()), data.getSize())) {
        DownloadGameAssetsFileTask task = new DownloadGameAssetsFileTask(launcher, data, progress);
        ThreadUtils.startThread(task);
      } else {
        progress.countDown();
      }
    }
    progress.await();
  }

  @EqualsAndHashCode
  @Getter
  public static class AssetsFileData {
    @SerializedName("hash")
    private String hash;
    @SerializedName("size")
    private int size;

    public String getHashSplicing() {
      return getHashHead() + "/" + hash;
    }

    public String getHashHead() {
      return hash.substring(0, 2);
    }
  }

  public static class DownloadGameAssetsFileTask extends Thread {
    private final Launcher launcher;
    private final AssetsFileData data;
    private final SpeedOfProgress progress;
    private final DownloadSourceInterface source;

    public DownloadGameAssetsFileTask(Launcher launcher, AssetsFileData data, SpeedOfProgress progress) {
      this.launcher = launcher;
      this.data = data;
      this.progress = progress;
      this.source = DownloadSourceManger.getDownloadSource();
    }

    @Override
    @SneakyThrows(IOException.class)
    public void run() {
      File HashFile = new File(launcher.getGameObjects(), data.getHashSplicing());
      URL HashUrl = new URL(source.getAssetsUrl() + data.getHashSplicing());
      DownloadUtils.StartDownloadTask(HashUrl, HashFile);
      synchronized (this) {
        progress.countDown();
      }
    }
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/download/game/DownloadGameAssetsFile.kt
package org.wdt.wdtc.core.download.game

import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.launch
import org.wdt.utils.gson.getJsonObject
import org.wdt.utils.gson.parseObject
import org.wdt.utils.gson.readFileToJsonObject
import org.wdt.wdtc.core.game.Version
import org.wdt.wdtc.core.manger.ProgressManger
import org.wdt.wdtc.core.manger.TaskKind
import org.wdt.wdtc.core.manger.TaskManger
import org.wdt.wdtc.core.manger.currentDownloadSource
import org.wdt.wdtc.core.utils.*
import java.io.File

class DownloadGameAssetsFile(private val version: Version) {
	suspend fun startDownloadAssetsFiles() {
		val assets = runOnIO {
			version.gameAssetsListJson
		}.readFileToJsonObject().getJsonObject("objects").asMap().values.map {
			it.asJsonObject.parseObject<AssetsFileData>()
		}
		
		val progress = ProgressManger(assets.size).apply {
			coroutineScope = executorCoroutineScope(name = "Download assets file")
		}
		assets.forEach { data ->
			DownloadGameAssetsFileTask(version, data, progress).run {
				start()
			}
		}
		progress.await()
	}
	
	class AssetsFileData(
		@SerializedName("hash")
		override val sha1: String,
		@SerializedName("size")
		override val size: Long = 0
	) : FileData {
		val hashSplicing: String
			get() = "$hashHead/$sha1"
		private val hashHead: String
			get() = sha1.substring(0, 2)
	}
	
	private class DownloadGameAssetsFileTask(
		version: Version, private val data: AssetsFileData, private val progress: ProgressManger
	) : TaskManger(data.sha1, TaskKind.COROUTINES) {
		
		private val hashFile = File(version.gameObjects, data.hashSplicing)
		
		init {
			coroutinesAction = progress.run {
				coroutineScope.launch(actionName.toCoroutineName(), CoroutineStart.LAZY) {
					startDownloadTask((currentDownloadSource.assetsUrl + data.hashSplicing).toURL() to hashFile)
					countDown()
				}
			}
		}
		
		override fun start() {
			if (hashFile compareFile data) {
				coroutinesAction.noNull().start()
			} else {
				progress.countDown()
			}
		}
	}
}
